rasa/core/featurizers/tracker_featurizers.py

```diff
@@ -0,0 +1,356 @@
+import jsonpickle
+import logging
+import os
+from tqdm import tqdm
+from typing import Tuple, List, Optional, Dict, Text, Deque, FrozenSet
+import numpy as np
+
+import rasa.utils.io
+from rasa.core.featurizers.single_state_featurizer import SingleStateFeaturizer
+from rasa.core.domain import Domain, State
+from rasa.core.events import ActionExecuted
+from rasa.core.trackers import DialogueStateTracker, FrozenState
+from rasa.utils.common import is_logging_disabled
+from rasa.utils.features import Features
+from rasa.core.interpreter import NaturalLanguageInterpreter
+from rasa.core.constants import USER
+from rasa.nlu.constants import TEXT
+
+logger = logging.getLogger(__name__)
+
+
+class TrackerFeaturizer:
+    """Base class for actual tracker featurizers."""
+
+    def __init__(
+        self, state_featurizer: Optional[SingleStateFeaturizer] = None
+    ) -> None:
+
+        self.state_featurizer = state_featurizer
+
+    @staticmethod
+    def _unfreeze_states(states: Deque[FrozenState]) -> List[State]:
+        return [
+            {key: dict(value) for key, value in dict(state).items()} for state in states
+        ]
+
+    def _create_states(
+        self, tracker: DialogueStateTracker, domain: Domain
+    ) -> List[State]:
+        """Create states: a list of dictionaries."""
+
+        states = tracker.past_states(domain)
+
+        return self._unfreeze_states(states)
```

Comment:
it seems counterintuitive to freeze and then unfreeze the states in the next step?

Reply:
yes, but otherwise will need to create several additional methods

Reply:
because during training `tracker.past_states` returns already precomputed during generation frozenstate, it saves a lot of compute time

Reply:
>  returns already precomputed during generation frozenstate

What do you mean? I don't see where the `freeze_current_state` part is cached. 

> create several additional methods

Exactly one more method, correct?

Reply:
refactored